### PR TITLE
[nnc] Fuse loops that have variable bounds

### PR DESCRIPTION
**Pull request by navahgar**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #74658
* __->__ #74346

When loops have variables as their bounds, loop fusion did not work since it was checking for constant bounds. That check is unnecessary. All we need to check is if the bounds are equal.

Differential Revision: [D34942360](https://our.internmc.facebook.com/intern/diff/D34942360)

### CODE REVIEW

**Comment by navahgar** [comment]
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34942360).

**Comment by navahgar** [comment]
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34942360).

**Comment by navahgar** [comment]
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34942360).

File: torch/csrc/jit/tensorexpr/bounds_overlap.cpp [hunk 1]
```diff
@@ -225,8 +225,20 @@ std::vector<IndexBounds> subtractIndicesBounds(
...
+        // In some cases, we might end up with empty remainingSlices due to the
+        // optimization done in subtraction while handling diff expressions
+        // that have a single variable in `subtractBound()`.
+        //
+        // For example:
+        //    (j, j) - (j + v, j + v) => (j, j + v - 1)
+        // When we try to extract the remaining slices from (j, j), we perform:
+        //    (j, j) - (j, j + v - 1) => ()
+        // and end up with empty remainingSlices.
+        if (!remainingSlices.empty()) {
```
Comment (eellison) [inline torch/csrc/jit/tensorexpr/bounds_overlap.cpp:237]: @ZolotukhinM mind taking a look at this part?

Comment (eellison) [reply torch/csrc/jit/tensorexpr/bounds_overlap.cpp:237]: bump @ZolotukhinM when you get a chance

Comment (ZolotukhinM) [reply torch/csrc/jit/tensorexpr/bounds_overlap.cpp:237]: @navahgar tbh, I have difficult times understanding what this code does (the entire function). I know it's not your code, but if you understand it, would you mind adding some comments to it?

To name some:
1) What does the third argument `overlap` do? Why is it needed? How does it control the behavior of the function and why?
2) What are these slices and how we end up getting them? In your comment it was not clear to me how from `(j, j) - (j + v, j + v)` we got `(j, j + v - 1)` and why do we then subtract it again from `(j, j)`? Btw, does this notation mean that we're trying to subtract a range with `start=end=j+v` from a range with `start=end=j`?

**Comment by navahgar** [comment]
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34942360).

File: torch/csrc/jit/tensorexpr/bounds_overlap.cpp (hunk 1 above)
Comment (navahgar) [reply torch/csrc/jit/tensorexpr/bounds_overlap.cpp:231]: > I have difficult times understanding what this code does (the entire function). I know it's not your code, but if you understand it, would you mind adding some comments to it?

I have added some explanation for the function in `torch/csrc/jit/tensorexpr/bounds_overlap.h`. PTAL

> What does the third argument overlap do? Why is it needed? How does it control the behavior of the function and why?

It was a mistake that `overlap` handling was done in the caller and not in this function. Fixed that as well.

This particular check is only for cases where the slices are empty during processing. The explanation I added there was confusing. So, removed it.

File: torch/csrc/jit/tensorexpr/bounds_overlap.cpp [hunk 2]
```diff
@@ -114,6 +114,13 @@ OverlapKind overlaps(const IndexBounds& a, const IndexBounds& b) {
 }
 
 std::vector<Bound> subtractBound(Bound a, Bound b, OverlapKind overlap) {
+  if (overlap == NoOverlap) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/bounds_overlap.cpp:117]: Do we even need to have these two versions of the function? Why don't we inline this one into the second one? Do we expect that the version with the "overlap" argument is independently useful too? In other words, isn't overlap handling an internal detail that should not be exposed in the API?

File: torch/csrc/jit/tensorexpr/bounds_overlap.h [hunk 3]
```diff
@@ -78,6 +78,14 @@ OverlapKind TORCH_API overlaps(const IndexBounds& a, const IndexBounds& b);
 // Multiple Bounds can be returned in the case where B slices A into two
 // distinct regions with no overlap.
 //
+// For example:
+//    subtractBound((0, 10), (2, 4)) => [(0, 1), (5, 10)]
+//       bound A: (0, 10)
+//       bound B: (2, 4)
+//       If we remove slice (2, 4) from the slice (0, 10), we will be left
+//       with 2 slices, one at the start (0, 1), and one at the end (5, 10).
+//       So, the result of this subtraction is [(0, 1), (5, 10)].
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/bounds_overlap.h:87]: That's super helpful! I realized that I completely misunderstood the function, I thought we were figuring out a boundary for `i - j` expression given boundaries for `i` and `j`.

**Comment by navahgar** [comment]
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34942360).

File: torch/csrc/jit/tensorexpr/bounds_overlap.cpp (hunk 2 above)
Comment (navahgar) [reply torch/csrc/jit/tensorexpr/bounds_overlap.cpp:118]: Sure, good point. Removed the other overload.

**Review by ZolotukhinM — APPROVED** [review]
The changes look good to me!

**Comment by navahgar** [comment]
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34942360).

**Pull request closed without merging**